aws-sdk-core/lib/aws-sdk-core/shared_config.rb

```diff
@@ -0,0 +1,179 @@
+
+    # @return [String]
+    attr_reader :config_path
+
+    # @return [String]
+    attr_reader :profile_name
+
+    # Constructs a new SharedConfig provider object. This will load the shared
+    # credentials file, and optionally the shared configuration file, as ini
+    # files which support profiles.
+    #
+    # By default, only the shared credential file (the default path for which is
+    # `~/.aws/credentials`) is loaded. However, if you set the
+    # `ENV['AWS_SDK_LOAD_CONFIG']` environment variable, the shared config file
+    # will also be loaded (the default path for which is `~/.aws/config`).
+    #
+    # The default profile name is 'default'. You can specify the profile name
+    # with the `ENV['AWS_PROFILE']` environment variable or with the
+    # `:profile_name` option.
+    #
+    # @param [Hash] options
+    # @option options [String] :credentials_path Path to the shared credentials
+    #   file. Defaults to "#{Dir.home}/.aws/credentials".
+    # @option options [String] :config_path Path to the shared config file.
+    #   Defaults to "#{Dir.home}/.aws/config".
+    # @option options [String] :profile_name The credential/config profile name
+    #   to use. If not specified, will check `ENV['AWS_PROFILE']` before using
+    #   the fixed default value of 'default'.
+    def initialize(options = {})
+      @profile_name = determine_profile(options)
+      @credentials_path = options[:credentials_path] ||
+        determine_credentials_path
+      load_credentials_file if loadable?(@credentials_path)
+      if ENV["AWS_SDK_LOAD_CONFIG"]
+        @config_path = options[:config_path] || determine_config_path
+        load_config_file if loadable?(@config_path)
+      end
+    end
+
+    # @api private
+    def fresh(options = {})
+      @profile_name = nil
+      @credentials_path = nil
+      @config_path = nil
+      @parsed_credentials = nil
+      @parsed_config = nil
+      @profile_name = determine_profile(options)
+      @credentials_path = options[:credentials_path] ||
+        determine_credentials_path
+      load_credentials_file if loadable?(@credentials_path)
+      if ENV["AWS_SDK_LOAD_CONFIG"]
+        @config_path = options[:config_path] || determine_config_path
+        load_config_file if loadable?(@config_path)
+      end
+    end
+
+    # @return [Boolean] Returns `true` if a credential file
+    #   exists and has appropriate read permissions at {#path}.
+    # @note This method does not indicate if the file found at {#path}
+    #   will be parsable, only if it can be read.
+    def loadable?(path)
+      !path.nil? && File.exist?(path) && File.readable?(path)
+    end
+
+    def credentials(opts = {})
+      p = opts[:profile] || @profile_name
+      validate_profile_exists(p) if @parsed_credentials || @parsed_config
+      if credentials = credentials_from_shared(p, opts)
+        credentials
+      elsif credentials = credentials_from_config(p, opts)
+        credentials
+      else
+        nil
```

Comment:
I've added some specific tests and configuration cases around this behavior. It will now explicitly fall through shared credentials, then to shared config, then to returning `nil` if credentials are incomplete as defined by the `Aws::Credentials#set?` method.